src/utils/AdaptiveTimeStep.cpp

```diff
@@ -116,24 +132,39 @@ AdaptiveTimeStep::Calculate (amrex::Real& dt, MultiBeam& beams, amrex::Real plas
         const auto uzp = soa.GetRealData(BeamIdx::uz).data() + box_offset;
         const auto wp = soa.GetRealData(BeamIdx::w).data() + box_offset;
 
-
         amrex::ReduceOps<amrex::ReduceOpSum, amrex::ReduceOpSum,
-                         amrex::ReduceOpSum, amrex::ReduceOpMin> reduce_op;
-        amrex::ReduceData<amrex::Real, amrex::Real,
+                         amrex::ReduceOpSum, amrex::ReduceOpMin,
+                         amrex::ReduceOpMin> reduce_op;
+        amrex::ReduceData<amrex::Real, amrex::Real, amrex::Real,
                           amrex::Real, amrex::Real> reduce_data(reduce_op);
         using ReduceTuple = typename decltype(reduce_data)::Type;
 
+        // Data required to gather the Ez field
+        const amrex::FArrayBox& slice_fab = fields.getSlices(lev)[0];
+        Array3<const amrex::Real> const slice_arr = slice_fab.const_array();
+        const int ez_comp = Comps[WhichSlice::This]["Ez"];
+        amrex::Real const * AMREX_RESTRICT dx = geom.CellSize();
+        const amrex::Real dx_inv = 1._rt/dx[0];
+        const amrex::Real dy_inv = 1._rt/dx[1];
+        amrex::Real const x_pos_offset = GetPosOffset(0, geom, slice_fab.box());
+        const amrex::Real y_pos_offset = GetPosOffset(1, geom, slice_fab.box());
+
         reduce_op.eval(numParticleOnTile, reduce_data,
             [=] AMREX_GPU_DEVICE (long ip) noexcept -> ReduceTuple
             {
                 if (pos_structs[ip].id() < 0) return {
-                    0._rt, 0._rt, 0._rt, std::numeric_limits<amrex::Real>::infinity()
+                    0._rt, 0._rt, 0._rt, std::numeric_limits<amrex::Real>::infinity(), 0._rt
                 };
+                amrex::Real Ezp = 0._rt;
+                doGatherEz(pos_structs[ip].pos(0), pos_structs[ip].pos(1), Ezp, slice_arr, ez_comp,
+                           dx_inv, dy_inv, x_pos_offset, y_pos_offset);
+                const amrex::Real uz = uzp[ip] + dt * charge_mass_ratio * Ezp;
```

Comment:
Could you explain this line? Instead of using the current `uz` you propose to use the next `uz`? I don't think this should be the case, e.g. when determining the initial time step.